hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/AbstractSerializationService.java

```diff
@@ -0,0 +1,466 @@
+        if (!(object instanceof Data)) {
+            return (T) object;
+        }
+
+        Data data = (Data) object;
+        if (isNullData(data)) {
+            return null;
+        }
+
+        BufferPool pool = bufferPoolThreadLocal.get();
+        BufferObjectDataInput in = pool.takeInputBuffer(data);
+        try {
+            final int typeId = data.getType();
+            final SerializerAdapter serializer = serializerFor(typeId);
+            if (serializer == null) {
+                if (active) {
+                    throw new HazelcastSerializationException("There is no suitable de-serializer for type " + typeId);
+                }
+                throw new HazelcastInstanceNotActiveException();
+            }
+
+            Object obj = serializer.read(in);
+            if (managedContext != null) {
+                obj = managedContext.initialize(obj);
+            }
+            return (T) obj;
+        } catch (Throwable e) {
+            throw handleException(e);
+        } finally {
+            pool.returnInputBuffer(in);
+        }
+    }
+
+    @Override
+    public final void writeObject(final ObjectDataOutput out, final Object obj) {
+        if (obj instanceof Data) {
+            throw new HazelcastSerializationException(
+                    "Cannot write a Data instance! " + "Use #writeData(ObjectDataOutput out, Data data) instead.");
+        }
+        try {
+            SerializerAdapter serializer = serializerFor(obj);
+            out.writeInt(serializer.getTypeId());
+            serializer.write(out, obj);
+        } catch (Throwable e) {
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public final <T> T readObject(final ObjectDataInput in) {
+        try {
+            final int typeId = in.readInt();
+            final SerializerAdapter serializer = serializerFor(typeId);
+            if (serializer == null) {
+                if (active) {
+                    throw new HazelcastSerializationException("There is no suitable de-serializer for type " + typeId);
+                }
+                throw new HazelcastInstanceNotActiveException();
+            }
+            Object obj = serializer.read(in);
+            if (managedContext != null) {
+                obj = managedContext.initialize(obj);
+            }
+            return (T) obj;
+        } catch (Throwable e) {
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public final void writeData(ObjectDataOutput out, Data data) {
+        try {
+            boolean isNull = data == null;
+            out.writeBoolean(isNull);
+            if (isNull) {
+                return;
+            }
+            writeDataInternal(out, data);
+        } catch (Throwable e) {
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public Data readData(ObjectDataInput input) {
+        return SerializationServiceHelper.readData(input);
```

Comment:
`readData` and `writeData` looks odd. One is in SerializationServiceHelper and the other is in AbstractSerializationService ?!?!